jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/compress/CompressExtension.java

```diff
@@ -18,350 +18,312 @@
+                catch (DataFormatException e)
+                {
+                    throw new BadPayloadException(e);
+                }
+            }
+        };
     }
 
-    /**
-     * Indicates use of RSV1 flag for indicating deflation is in use.
-     */
     @Override
     public boolean isRsv1User()
     {
         return true;
     }
 
-    /**
-     * Return the mode of operation for dropping (or keeping) tail bytes in frames generated by compress (outgoing)
-     *
-     * @return either {@link #TAIL_DROP_ALWAYS}, {@link #TAIL_DROP_FIN_ONLY}, or {@link #TAIL_DROP_NEVER}
-     */
-    abstract int getTailDropMode();
-
-    /**
-     * Return the mode of operation for RSV flag use in frames generate by compress (outgoing)
-     *
-     * @return either {@link #RSV_USE_ALWAYS} or {@link #RSV_USE_ONLY_FIRST}
-     */
-    abstract int getRsvUseMode();
-
-    protected void forwardIncoming(Frame frame, Callback callback, ByteAccumulator accumulator)
-    {
-        Frame newFrame = Frame.copyWithoutPayload(frame);
-
-        // Unset RSV1 since it's not compressed anymore.
-        newFrame.setRsv1(false);
+    abstract CompressionMode getCompressionMode();
 
-        ByteBuffer buffer = getBufferPool().acquire(accumulator.getLength(), false);
-        try
-        {
-            int position = BufferUtil.flipToFill(buffer);
-            try
-            {
-                accumulator.transferTo(buffer);
-            }
-            finally
-            {
-                BufferUtil.flipToFlush(buffer, position);
-            }
+    @Override
+    public void sendFrame(Frame frame, Callback callback, boolean batch)
+    {
+        // Compressed frames may increase in size so we need the flusher to fragment them.
+        outgoingFlusher.sendFrame(frame, callback, batch);
+    }
 
-            newFrame.setPayload(buffer);
-            nextIncomingFrame(newFrame, callback);
-        }
-        finally
-        {
-            getBufferPool().release(buffer);
-        }
+    @Override
+    public void onFrame(Frame frame, Callback callback)
+    {
+        incomingFlusher.sendFrame(frame, callback, false);
     }
 
-    protected void decompress(ByteAccumulator accumulator, ByteBuffer buf) throws DataFormatException
+    protected boolean decompress(Frame frame, Callback callback, boolean first) throws DataFormatException
     {
-        if ((buf == null) || (!buf.hasRemaining()))
-        {
-            return;
-        }
-        byte[] output = new byte[DECOMPRESS_BUF_SIZE];
+        ByteBuffer data = frame.getPayload();
+        long maxFrameSize = getWebSocketCoreSession().getMaxFrameSize();
 
         Inflater inflater = getInflater();
+        if (first)
+            inflater.setInput(data);
 
-        while (buf.hasRemaining() && inflater.needsInput())
+        // Decompress the payload.
+        boolean finished = false;
+        ByteAccumulator accumulator = new ByteAccumulator(getWebSocketCoreSession().getMaxFrameSize());
```

Comment:
why do we accumulate?

Reply:
We're not accumulating multiple frames, we're just getting `DECOMPRESS_BUF_SIZE` chunks of the inflated data until we reach the end of data or get to `maxFrameSize`. Then we move all these data chunks to a buffer of the correct size and send that as the frame payload.

This is the same behaviour that existed before but with one less copy of the data, this PR is just implementing autoFragmentation.

Reply:
We should never accumulate if Auto fragment is true

Reply:
To me the issue of accumulating is precisely the issue of fragmenting.  If we can fragment, then we never need to accumulate. The only reason we'd even need to accumulate is if we were not allowed to fragment... but then if we exceed the buffer size, then that could be configured to maxFrame and we are not able to do anything anyway.

No accumulation please!